nav2_toolkit/src/pose_saver_node.cpp

```diff
@@ -0,0 +1,296 @@
+  }
+}
+
+void PoseSaverNode::write_pose_to_file(const std::string &filepath)
+{
+  if (!last_pose_) return;
+
+  YAML::Emitter out;
+  out << YAML::BeginMap;
+  out << YAML::Key << "pose" << YAML::BeginMap;
+  out << YAML::Key << "position" << YAML::BeginMap
+      << YAML::Key << "x" << YAML::Value << last_pose_->pose.pose.position.x
+      << YAML::Key << "y" << YAML::Value << last_pose_->pose.pose.position.y
+      << YAML::Key << "z" << YAML::Value << last_pose_->pose.pose.position.z
+      << YAML::EndMap;
+  out << YAML::Key << "orientation" << YAML::BeginMap
+      << YAML::Key << "x" << YAML::Value << last_pose_->pose.pose.orientation.x
+      << YAML::Key << "y" << YAML::Value << last_pose_->pose.pose.orientation.y
+      << YAML::Key << "z" << YAML::Value << last_pose_->pose.pose.orientation.z
+      << YAML::Key << "w" << YAML::Value << last_pose_->pose.pose.orientation.w
+      << YAML::EndMap;
+  out << YAML::EndMap;
+  out << YAML::EndMap;
+
+  std::string tmp_path = filepath + ".tmp";
+  std::ofstream tmp_out(tmp_path);
+  tmp_out << out.c_str();
+  tmp_out.close();
+
+  std::rename(tmp_path.c_str(), filepath.c_str());
+}
+void PoseSaverNode::amcl_monitor_callback()
+{
+  if (!auto_restore_) return;
+
+  int current_count = initial_pose_pub_->get_subscription_count();
+  if (last_sub_count_ > 0 && current_count == 0) {
+    RCLCPP_WARN(this->get_logger(), "Lost AMCL. Waiting...");
+  }
+  if (last_sub_count_ == 0 && current_count > 0) {
+    RCLCPP_INFO(this->get_logger(), "AMCL reconnected! Publishing last known pose.");
+    try {
+      auto pose_msg = read_pose_from_file(pose_file_path_);
+      rclcpp::sleep_for(500ms);
+      initial_pose_pub_->publish(pose_msg);
+    } catch (const std::exception &e) {
+      RCLCPP_WARN(this->get_logger(), "Failed to auto-restore after AMCL reconnect: %s", e.what());
+    }
+  }
+  last_sub_count_ = current_count;
+}
+
+
+geometry_msgs::msg::PoseWithCovarianceStamped PoseSaverNode::read_pose_from_file(const std::string &filepath)
+{
+  YAML::Node node = YAML::LoadFile(filepath);
+  geometry_msgs::msg::PoseWithCovarianceStamped msg;
+  msg.header.stamp = this->now();
+  msg.header.frame_id = "map";
+  msg.pose.pose.position.x = node["pose"]["position"]["x"].as<double>();
+  msg.pose.pose.position.y = node["pose"]["position"]["y"].as<double>();
+  msg.pose.pose.position.z = node["pose"]["position"]["z"].as<double>();
+  msg.pose.pose.orientation.x = node["pose"]["orientation"]["x"].as<double>();
+  msg.pose.pose.orientation.y = node["pose"]["orientation"]["y"].as<double>();
+  msg.pose.pose.orientation.z = node["pose"]["orientation"]["z"].as<double>();
+  msg.pose.pose.orientation.w = node["pose"]["orientation"]["w"].as<double>();
+  return msg;
+}
+
+std::string PoseSaverNode::get_package_config_path()
+{
+  return ament_index_cpp::get_package_share_directory("nav2_toolkit") + "/config/last_known_pose.yaml";
+}
+
+}  // namespace nav2_toolkit
+
+int main(int argc, char **argv)
```

Comment:
Move the main into another file so that this node can be used as a component. 

Also, make this node a component node :-)